packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts

```diff
@@ -806,12 +808,31 @@ export class ModularEditBuilder
 	extends ProgressiveEditBuilderBase<ModularChangeset>
 	implements ProgressiveEditBuilder<ModularChangeset>
 {
+	private transactionDepth: number = 0;
+	private idAllocator: IdAllocator;
+
 	public constructor(
-		family: ChangeFamily<unknown, ModularChangeset>,
+		family: ChangeFamily<ChangeFamilyEditor, ModularChangeset>,
 		changeReceiver: (change: ModularChangeset) => void,
 		anchors: AnchorSet,
 	) {
 		super(family, changeReceiver, anchors);
+		this.idAllocator = () => brand(0);
+	}
+
+	public override enterTransaction(): void {
+		this.transactionDepth += 1;
+		if (this.transactionDepth === 1) {
+			this.idAllocator = idAllocatorFromMaxId();
+		}
+	}
+
+	public override exitTransaction(): void {
+		assert(this.transactionDepth > 0, "Cannot exit inexistent transaction");
+		this.transactionDepth -= 1;
+		if (this.transactionDepth === 0) {
+			this.idAllocator = () => brand(0);
```

Comment:
It seems like this assumes that an edit will never need to allocate more than one ID.

Reply:
Good point. That doesn't seem very safe.
I see 3 options:
1. Make callers of `generateId()` reponsible for calling enter/exit transaction when they need to call `generateId()` multiple times.
2. Have the non-transaction state just be `this.idAllocator = idAllocatorFromMaxId()` instead of `this.idAllocator = () => brand(0)`.
3. Leave it as is. Document the assumption on the `generateId()` method.

Option 1 seems to complicate the contract in an error-prone way.
Option 2 has the disavantage of leading to ever-increasing move IDs when transactions are not used. Which leads to a little bit of bloat in changesets.
Option 3 is also a bit ugly contract-wise, and it easy to mess up.

I'm going to go for option 2, but if you can think of something else, let me know.